ast/check.go

```diff
@@ -61,124 +81,211 @@ func (tc *typeChecker) CheckBody(env *TypeEnv, body Body) (*TypeEnv, Errors) {
 	return env, tc.errs
 }
 
-// CheckRules runs type checking on the rules and returns a TypeEnv if no
-// errors are found. The resulting TypeEnv wraps the provided one. The
-// resulting TypeEnv will be able to resolve types of refs that refer to rules.
-func (tc *typeChecker) CheckRules(env *TypeEnv, rules []*Rule) (*TypeEnv, Errors) {
+func (tc *typeChecker) checkLanguageBuiltins(env *TypeEnv) {
+	for _, bi := range Builtins {
```

Comment:
Nit: I would expect the type info for these built-ins to be inserted into their own (wrapped) type env and not merged with the type env containing info for rules and user funcs.